sw/host/opentitanlib/BUILD

```diff
@@ -231,6 +234,8 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
+        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
```

Comment:
why is this necessary?

Reply:
Thanks for asking.  What happened when I wrote `"openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)"` was that bazel complained when compiling `opentitanlib_test` that it was referring to the data dependency through `$(rootpath ...)`, but did not itself declare it.  I could make that error go away by duplicating the `data = [ ... ]` declaration in `opentitanlib_test`.  However, next thing was that bazel complained that `opentitanlib_doc` also referred to the data dependency without declaring it.  It does not seem right to duplicate the `data` that way (and `rust_doc` does not allow `data` anyway.)

If anyone knows of a better way of referring to a data files from a library like this, I am happy to make the change.

Reply:
I will have a look at this.

Reply:
I was able to make it work with:
```diff
diff --git a/sw/host/opentitanlib/BUILD b/sw/host/opentitanlib/BUILD
index e2b6610d1..40a93cecb 100644
--- a/sw/host/opentitanlib/BUILD
+++ b/sw/host/opentitanlib/BUILD
@@ -213,6 +213,7 @@ rust_library(
         ":pinmux_config",
         ":spi_passthru",
         "@hyperdebug_firmware//:hyperdebug/ec.bin",
+        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
     ],
     crate_features = [
         "include_hyperdebug_firmware",
@@ -220,9 +221,6 @@ rust_library(
         "//sw/device:is_english_breakfast": ["english_breakfast"],
         "//conditions:default": [],
     }),
-    data = [
-        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
-    ],
     proc_macro_deps = [
         "//sw/host/opentitanlib/opentitantool_derive",
         "@lowrisc_serde_annotate//annotate_derive:annotate_derive",
@@ -234,8 +232,7 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
-        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
-        "openocd_cmsis_dap_adapter_cfg": "external/openocd/tcl/interface/jtag_cmsis_dap_adapter_cfg",
+        "openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)",
     },
     deps = [
         "//hw/top_earlgrey/sw/autogen/chip:top_earlgrey",
```
I think the problem is that you put it in `data` instead of `compile_data`.

Reply:
What do you mean by "make it work".  Did you attempt running the code and see opentitantool being able to spawn openocd, and openocd accessing the file?

As far as I understand, `compile_data`, as its name suggests, is only available at compile time, and not later at runtime, when this file is going to be needed for an openocd sub-process.  (Maybe I could do something funny, and include the data in the binary by `include_str!()`, and start the openocd sub-process with something like `/dev/fd/17` on the command line as the configuration file, and create anonymous pipes to send the data through.  This might work, but it is also possible that openocd wants to `seek()` in the file, or do something else that does not work on pipes.  And I do not know if the Rust sub-process library allows the creation of such anonymous pipes besides stdin, stdout and stderr.  I do not want to go there.)

Reply:
Ah sorry, I misunderstood what you are trying to achieve here. You are trying to embed the **path** (not content) to a bazel-managed file when you create the library. I should point out that this is an extremely fragile way of doing things because you are relying on a stable path at compile and execution time. For this you need to pass the file as both compile_data and data. Then you ca use rootpath that actually guarantees to use a workspace-relative path so it will be the same path at compile time and execution time.

```diff
diff --git a/sw/device/tests/BUILD b/sw/device/tests/BUILD
index ad2241c8c..f42eb7b56 100644
--- a/sw/device/tests/BUILD
+++ b/sw/device/tests/BUILD
@@ -2827,6 +2827,26 @@ opentitan_functest(
     ],
 )
 
+opentitan_functest(
+    name = "openocd_hyperdebug_test",
+    srcs = ["example_test_from_flash.c"],
+    cw310 = cw310_params(
+        interface = "hyper310",
+        bitstream = "//hw/bitstream/hyperdebug:rom_with_fake_keys_otp_test_unlocked0",
+        test_cmds = [
+            "--bitstream=\"$(rootpath {bitstream})\"",
+            "--bootstrap=\"$(rootpath {flash})\"",
+        ] + OPENTITANTOOL_OPENOCD_TEST_CMDS,
+    ),
+    data = OPENTITANTOOL_OPENOCD_DATA_DEPS,
+    targets = ["cw310_rom_with_fake_keys"],
+    test_harness = "//sw/host/tests/chip/jtag:openocd_test",
+    deps = [
+        "//sw/device/lib/runtime:log",
+        "//sw/device/lib/testing/test_framework:ottf_main",
+    ],
+)
+
 opentitan_functest(
     name = "spi_device_ottf_console_test",
     srcs = ["spi_device_ottf_console_test.c"],
diff --git a/sw/host/opentitanlib/BUILD b/sw/host/opentitanlib/BUILD
index e2b6610d1..c6870b338 100644
--- a/sw/host/opentitanlib/BUILD
+++ b/sw/host/opentitanlib/BUILD
@@ -213,6 +213,7 @@ rust_library(
         ":pinmux_config",
         ":spi_passthru",
         "@hyperdebug_firmware//:hyperdebug/ec.bin",
+        "//third_party/openocd:jtag_cmsis_dap_adapter_cfg",
     ],
     crate_features = [
         "include_hyperdebug_firmware",
@@ -234,8 +235,12 @@ rust_library(
         "pinmux_config": "$(location :pinmux_config)",
         "spi_passthru": "$(location :spi_passthru)",
         "hyperdebug_firmware": "$(location @hyperdebug_firmware//:hyperdebug/ec.bin)",
-        # Manual expansion of: $(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)
-        "openocd_cmsis_dap_adapter_cfg": "external/openocd/tcl/interface/jtag_cmsis_dap_adapter_cfg",
+        # NOTE: here we must use rootpath, this is because we are using file as both compile_data
+        # to embed the path name, and as data to actually load it at runtime. For this purpose,
+        # the name must be independent of the execution environment's path. This guarantee is provided
+        # by rootpath that will generate a pure workspace-relative path:
+        # https://bazel.build/reference/be/make-variables#predefined_label_variables
+        "openocd_cmsis_dap_adapter_cfg": "$(rootpath //third_party/openocd:jtag_cmsis_dap_adapter_cfg)"
     },
     deps = [
         "//hw/top_earlgrey/sw/autogen/chip:top_earlgrey",
```
I actually tried it with
```bash
./bazelisk.sh test --test_output=streamed //sw/device/tests:openocd_hyperdebug_test
```
and it worked.

Reply:
Thank you, this helped immensely.

I have done the changes as you have suggested.  I have also added another CL into this PR, which refactors how configuration files for the existing Olimex JTAG adapter is passed into Rust code.  I think it is a simplification, let me know what you think.

I could back out the refactor, and instead add one more JTAG command line argument for the HyperDebug adapter, though it does not feel right to me that the caller of `opentitantool`, who already specify `--interface cw310` or `--interface hyper310` to have to also know details about which OpenOCD configuration file each of the backends need.

Reply:
I agree with trying to keep opentitanlib separate from Bazel at runtime, and embedding the config sounds sensible. Writing it to a temporary file sounds fiddly, but maybe there’s a way to pass config through the CLI?

Reply:
I think you can pass the entire content of openocd config files on the command line with the `-c` switch, instead of the `-f` switch.  Let me look into that.

I agree that it could make sense to swap out the olimex configuration file, and hence that it should be passed as a parameter.  I do not think that it belongs in the `io::jtag::JtagParams`, which is part of the `Jtag` trait that is shared by all backend drivers, though, but rather should be a parameter to the `ChipWhisperer` constructor.  I am going to attempt such a refactor as well.

How about the openocd executable itself?  We currently depend on a patched version.  Should that be embedded into opentitanlib as well, so that at runtime it could write a temporary file, make it executable, and start it?  For "outside" users who do not use bazel, it probably makes sense that opentitantool attempts to use the local openocd from the users $PATH, (as long as we know that there is a chance it could work.)  Does anyone know the nature of the local changes to openocd? Could they be upstreamed such that we do not need to depend on patches?